File: Reverse_day_3/reverse.js
// Brute force approach

//  Time Complexity: O(n)
//  Space Complexity: O(n) 

function reverseArray(arr) {
    let reversedArr = [];

    for(let i = arr.length-1; i>=0; i--){
        reversedArr.push(arr[i]);
    }
    return reversedArr;
}


// optimized approach

//  Time Complexity: O(n)
//  Space Complexity: O(1)


function reverseArrayOptimized(arr) {
    let left = 0;
    let right = arr.length - 1;

    while(left < right){
        [arr[left], arr[right]] = [arr[right], arr[left]];
        left++;
        right--;
    }
    return arr;
}

File: Reverse_day_3/reverse.py
# Brute force solution
arr = [1, 2, 3, 4, 5]
def reverse_string(s):
    reversed = []
    for i in range(len(s) - 1, -1, -1):
        reversed.append(s[i])
    return reversed
result = reverse_string(arr)
print(result)  


# Optimized solution

def reverse_string_optimized(s):
    left = 0
    right = len(s) - 1
    while left < right:
        s[left], s[right] = s[right], s[left]
        left += 1
        right -= 1
    return s

result = reverse_string_optimized(arr)
print(result)  
